system/lib/pthread/library_pthread.c

```diff
@@ -214,10 +231,52 @@ static void _do_call(em_queued_call *q)
 }
 
 #define CALL_QUEUE_SIZE 128
-static em_queued_call **call_queue = 0;
-static volatile int call_queue_head = 0; // Shared data synchronized by call_queue_lock.
-static volatile int call_queue_tail = 0;
+
+typedef struct CallQueue
+{
+	void *target_thread;
+	em_queued_call **call_queue;
+	volatile int call_queue_head; // Shared data synchronized by call_queue_lock.
+	volatile int call_queue_tail;
+	volatile struct CallQueue *next;
+} CallQueue;
+
+// Currently global to the queue, but this can be improved to be per-queue specific. (TODO: with lockfree list operations on callQueue_head, or removing the list by moving this data to pthread_t)
 static pthread_mutex_t call_queue_lock = PTHREAD_MUTEX_INITIALIZER;
+static volatile CallQueue *callQueue_head = 0;
+
+static volatile CallQueue *GetQueue(void *target) // Not thread safe, call while having call_queue_lock obtained.
+{
+	assert(target);
+	volatile CallQueue *q = callQueue_head;
```

Comment:
Regardless of the above comment that we shouldn't use volatile, it seems odd to me that some of this code just uses volatile pointers directly, and other (below) explicitly uses `emscripten_atomic_load_u32`. Is there some reason for the difference?

Reply:
The `emscripten_atomic_load_u32()` calls here were kind of looking forwards to a lock free linked list where we would not be having a `pthread_mutex_lock()` - but did not get around to doing that. Things should be safe here with the current `pthread_mutex_lock()` in place, and the only meaning of `volatile` here should be taken to be "observe all reads and writes", but nothing more. In wasm things should not be hurt with the `volatile`, even if they are implemented as atomic accesses. (fastcomp does require volatile accesses to be aligned, because it hoists them to atomics to avoid JS engines not having proper semantics for this)

Reply:
My point above is that the semantics of (non-atomic) volatile mean that reads and writes *are not* guaranteed to be observed by another thread at all. Only atomic makes any guarantees about that. In the absence of guarantees about observations from other threads, the possible observers that volatile makes guarantees about include the same thread (which is of course already covered by the "as-if" rule), as well as async signal handlers and memory-mapped devices. So in the absence of async signals and external devices, marking a load as volatile makes no additional guarantees beyond what the "as-if" rule does. (At least, that is Dan's interpretation as expressed in https://github.com/WebAssembly/tool-conventions/issues/60 *. That is why there is the possibility that volatiles will be implemented as non-atomic reads and writes). So in particular, reads and write to `callQueue_head` are not guaranteed to ever be observed by other threads. Because we do want that guarantee, we should use atomics (either C11 atomics or emscripten atomics) instead of volatiles here.

(*) There is some disagreement about what exactly is guaranteed by volatile because the spec lacks the language to really say anything about optimizations or observability outside of the C environment. But the C standard is clear that it's atomic operations that do make any guarantees about being observed by other threads.

Reply:
Originally the variables were marked volatile to highlight the shared state boundary that threads are working on, and to avoid a Firefox asm.js multithreading JIT bug. They were not there to provide inter-thread synchronization primitive. Anyhow, there have been too many what-is-volatile conversations, so removed the volatiles in this PR to avoid the issue.